src/diagnostics.rs

```diff
@@ -157,6 +157,46 @@ pub fn report_error<'tcx, 'mir>(
         }
     };
 
+    let mut stacktrace = ecx.generate_stacktrace();
+    let has_local_frame = stacktrace.iter().any(|frame| frame.instance.def_id().is_local());
+    match ecx.machine.backtrace_style {
+        BacktraceStyle::Off => {
+            // Retain one frame so that we can print a span for the error itself
+            stacktrace.truncate(1);
+        }
+        BacktraceStyle::Short => {
+            // Only prune frames if there is at least one local frame. This check ensures that if
+            // we get a backtrace that never makes it to the user code because it has detected a
+            // bug in the Rust runtime, we don't prune away every frame.
+            if has_local_frame {
+                // This is part of the logic that `std` uses to select the relevant part of a
+                // backtrace. But here, we only look for __rust_begin_short_backtrace, not
+                // __rust_end_short_backtrace because the end symbol comes from a call to the default
+                // panic handler.
+                stacktrace = stacktrace
+                    .into_iter()
+                    .take_while(|frame| {
+                        let def_id = frame.instance.def_id();
+                        let path = ecx.tcx.tcx.def_path(def_id).to_string_no_crate_verbose();
```

Comment:
I think we should be using https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.def_path_str

Reply:
Is it possible to get the `Ident` for the name of the item, and just look at that? Then we could even compare for equality with `__rust_begin_short_backtrace` instead of doing a substring match.

Reply:
I guess we havehttps://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.item_name and `def_span` or `def_ident_span`

Reply:
`TyCtxt::item_name` panics when I feed it this `DefId` which appears in `cargo miri test` backtraces: `DefId(0:8 ~ scratch[c101]::tests::it_works::{closure#0})`.

To be clear, in the current setup `to_string_no_crate_verbose` seems to return `::__rust_begin_short_backtrace` in `cargo miri test`, and `::sys_common::backtrace::__rust_begin_short_backtrace` in `cargo miri run`. We _could_ check for either one of those, or check if the string ends with `::__rust_begin_short_backtrace`, but I'm not sure any of these options is on net more robust than the others.

Reply:
Whoops, missed that. Done.